1. Write a welcome program that asks for the name of students recursively and stores the names in a list. Then, print out a welcome message for each name in the list. Your output should look like this:

Enter the name of a student (Enter q/Q to stop): Mark  
Enter the name of a student (Enter q/Q to stop): David  
Enter the name of a student (Enter q/Q to stop): q  
Hello Mark, welcome to the course 5002.  
Hello David, welcome to the course 5002.

In [1]:
#Q1 answer
names=[]
stop = False
while not stop:
    name = input('Enter the name of a student (Enter q/Q to stop): ')
    if name not in ['q','Q']:
        names.append(name)
    else:
        stop = True
for n in names:
    print('Hello {}, welcome to the course 5002.'.format(n))

Enter the name of a student (Enter q/Q to stop): Mark
Enter the name of a student (Enter q/Q to stop): David
Enter the name of a student (Enter q/Q to stop): q
Hello Mark, welcome to the course 5002.
Hello David, welcome to the course 5002.


2. Given a list of random numbers (rn), pick out the ones that are multiple of 3, and print them in ascending order. You may find the function sorted(list) useful. Your output should look like this:

For list [1,5,3,7,0,25,6], the result is [0,3,6].

In [2]:
#Q2 answer
def multiple3(list):
    return sorted([i for i in list if i%3==0])

from random import randint
rn=[randint(1,1500) for _ in range (50)]

test = [1,5,3,7,0,25,6]
print(multiple3(test))
print(multiple3(rn))

[0, 3, 6]
[201, 309, 369, 447, 525, 636, 675, 738, 975, 984, 1020, 1191, 1305, 1335, 1374]


3. Write a program to let people guess a number (target=9). The input number should be between 0 and 30. If the number is correct, display congratulations; if the number is smaller or larger than the target, display a hint; if the number is out of range, display a warning. Your output should look like this:

Enter a number between 0 and 30: 40  
Number out of range! Try again: 2  
Number is too small, try again: 7  
Number is too small, try again: 11  
Number is too large, try again: 9  
You got it!

In [3]:
#Q3 answer
#This function is to check whether n is a number, with int or float
def check_number(n):
    try:
        n=int(n)
        return n
    except ValueError:
        try:
            n = float(n)
            return n
        except ValueError:
            return False

# This is the main function
def guess(target):
    correct=False
    n = input('Enter a number between 0 and 30: ')
    while not correct:
        while not check_number(n):
            n = input('Input is not number! Try again: ')
        cn = check_number(n)
        if cn<0 or cn>30:
            n = input('Number our of range! Try again: ')
        elif cn>target:
            n = input('Number is too large, try again: ')
        elif cn<target:
            n = input('Number is too small, try again: ')
        else:
            print('You got it!')
            correct = True
    
target=9
guess(target)

Enter a number between 0 and 30: 40
Number our of range! Try again: 2
Number is too small, try again: 7
Number is too small, try again: 11
Number is too large, try again: 9
You got it!


4. Write a program for a sandwich shop to take orders. Ask the customers if they want Cheese sandwich or Veggie sandwich. If they want Veggie sandwich, print 'Here is your Veggie Special sandwich~'. If they want Cheese, ask further if they want Cheddar or Manchego cheese. If they type something else, print 'No such sandwich...'. Your output should look like either of these:

Hi, welcome to our sandwich shop. Please select a sandwich.  
"c" for Cheese or "v" for Veggie Special: C  
"c" for Cheddar or "m" for Manchego: M  
Here is your Manchego Cheese sandwich~
***********************
Hi, welcome to our sandwich shop. Please select a sandwich.  
"c" for Cheese or "v" for Veggie Special: v  
Here is your Veggie Special sandwich~

In [5]:
#Q4 answer
print('Hi, welcome to our sandwich shop. Please select a sandwich.')
c_or_v = input('"c" for Cheese or "v" for Veggie Special: ')
if c_or_v in ['C','c']:
    c_or_m = input('"c" for Cheddar or "m" for Manchego: ')
    if c_or_m in ['C','c']:
        print("Here is your Cheddar Cheese sandwich~")
    elif c_or_m in ['M','m']:
        print("Here is your Manchego Cheese sandwich~")
    else:
        print("No such sandwich...")
elif c_or_v in ['V','v']:
    print("Here is your Veggie Special sandwich~")
else:
    print("No such sandwich...")

Hi, welcome to our sandwich shop. Please select a sandwich.
"c" for Cheese or "v" for Veggie Special: v
Here is your Veggie Special sandwich~


5. Write a what-to-wear program that asks for the weather (sunny, rainy, snowy, etc.) input, then print out clothes suggestions. If the weather is sunny, print "Wear a t-shirt and apply sunscreen"; if the weather is rainy, print "Bring an umbrella and wear boots"; if the weather is snowy, print "Wear a warm coat and hat"; if other words (e.g. @@@@) are input, print 'Sorry, not sure what to suggest for @@@@ weather...'. Use a while loop to keep asking for the weather input until one of the following (sunny, rainy, snowy) is given. Your output should look like this:

Enter the weather (sunny, rainy, snowy, etc.): sdf  
Sorry, not sure what to suggest for sdf weather...  
Enter the weather (sunny, rainy, snowy, etc.): sunny  
Wear a t-shirt and apply sunscreen

In [6]:
#Q5 answer
weather = input('Enter the weather (sunny, rainy, snowy, etc.): ')
while weather not in ['sunny','rainy','snowy']:
    print('Sorry, not sure what to suggest for {} weather'.format(weather))
    weather = input('Enter the weather (sunny, rainy, snowy, etc.): ')
if weather == 'sunny':
    print('Wear a t-shirt and apply sunscreen')
elif weather == 'rainy':
    print('Bring an umbrella and wear boots')
else:
    print('Wear a warm coat and hat')

Enter the weather (sunny, rainy, snowy, etc.): sdf
Sorry, not sure what to suggest for sdf weather
Enter the weather (sunny, rainy, snowy, etc.): sunny
Wear a t-shirt and apply sunscreen


6. Given a random matrix of size 10 by 10, check if there are zeros on the diagonal axis. Your output should print T for true or F for false. Now write a loop that executes this check for 100 times, and get the statistics about how many times the output is True and how many times the output is False. Print out the statistics. You may need the following code to generate a 2D matrix: `matrix=[[randint(-1,3) for i in range(m)] for j in range(m)]`  
Your output should look like this:

TTTTTFTTTTTTTTTTTTTFTTFFTTTTTTTTTTTTTFTTTTTTTTTTTTFTTTTTTFTTTTTFTFTFTTTTFTTTTTFFTTTFTTFTTTFFTTTTTTTF  
Among 100 trials, there are 82 True and 18 False.

In [23]:
# Q6 answer
from random import randint
n=100
m = 10
out = []
for a in range(n):
    matrix=[[randint(-1,3) for i in range(m)] for j in range(m)]
    z = 'F'
    for b in range(m):
        if matrix[b][b]==0:
            z = 'T'
    out.append(z)
    print(z,end='')
print()
print('Among {n} trials, there are {T} True and {F} False.'.format(n=n,T=out.count('T'),F=out.count('F')))
    

TTTTFTTTTTTTTTTTTTTTTTTTTFTTTTFTFTFTTTTFTTTTTTTTTTTFTTTTTTTTTTTTTTTFTTTTTTTFTTTTTFTTTTFTTTTTTTTTTFTT
Among 100 trials, there are 88 True and 12 False.


7. Goldbach's conjecture states that every even integer greater than 2 can be expressed as the sum of two primes. Write a program to test the Goldbach's conjecture for all the even numbers between 2 and 1000. List the even numbers on the left and the 2 prime numbers on the right. Your output should look like this:

4 = 2 + 2  
6 = 3 + 3  
8 = 3 + 5  
10 = 3 + 7  
12 = 5 + 7  
...  
100 = 3 + 97  
...

In [8]:
#Q7 answer

#generate a list containing prime from 2 to n first
n=1000
p_list=[]
for i in range(2,n+1):
    is_prime = True
    for j in range(2,i):
        if i%j == 0:
            is_prime = False
    if is_prime:
        p_list.append(i)

#test Goldbach's conjecture
for i in range(4,n+1,2):
    paired = False
    c = 0
    while not paired :
        p1 = p_list[c]
        for p in p_list:
            if p1+p == i:
                p2 = p
                paired = True
                break
        c+=1
    print('{} = {} + {}'.format(i,p1,p2))
    
        


4 = 2 + 2
6 = 3 + 3
8 = 3 + 5
10 = 3 + 7
12 = 5 + 7
14 = 3 + 11
16 = 3 + 13
18 = 5 + 13
20 = 3 + 17
22 = 3 + 19
24 = 5 + 19
26 = 3 + 23
28 = 5 + 23
30 = 7 + 23
32 = 3 + 29
34 = 3 + 31
36 = 5 + 31
38 = 7 + 31
40 = 3 + 37
42 = 5 + 37
44 = 3 + 41
46 = 3 + 43
48 = 5 + 43
50 = 3 + 47
52 = 5 + 47
54 = 7 + 47
56 = 3 + 53
58 = 5 + 53
60 = 7 + 53
62 = 3 + 59
64 = 3 + 61
66 = 5 + 61
68 = 7 + 61
70 = 3 + 67
72 = 5 + 67
74 = 3 + 71
76 = 3 + 73
78 = 5 + 73
80 = 7 + 73
82 = 3 + 79
84 = 5 + 79
86 = 3 + 83
88 = 5 + 83
90 = 7 + 83
92 = 3 + 89
94 = 5 + 89
96 = 7 + 89
98 = 19 + 79
100 = 3 + 97
102 = 5 + 97
104 = 3 + 101
106 = 3 + 103
108 = 5 + 103
110 = 3 + 107
112 = 3 + 109
114 = 5 + 109
116 = 3 + 113
118 = 5 + 113
120 = 7 + 113
122 = 13 + 109
124 = 11 + 113
126 = 13 + 113
128 = 19 + 109
130 = 3 + 127
132 = 5 + 127
134 = 3 + 131
136 = 5 + 131
138 = 7 + 131
140 = 3 + 137
142 = 3 + 139
144 = 5 + 139
146 = 7 + 139
148 = 11 + 137
150 = 11 + 139
152 = 3 + 149
154 = 3 + 151
156 = 5 + 151
158 = 7 + 151
160 = 3

8. Capitalize all words in a title entered, except: a, an, the, at, by, for, in, of, on, to, up, and, as, but, or, and nor. Your output should look like this:

Enter a title: Welcome to the world of data-driven modeling for master students offered by UST in a great and green campus on the hill  
Capitalized: Welcome to the World of Data-Driven Modeling for Master Students Offered by UST in a Great and Green Campus on the Hill

In [9]:
#Q8 answer
x_words = ['a','an','the','at','by','for','in','of','on','to','up','and','as','but','or','nor']
input_words = input('Enter a title: ')
split_space = input_words.split()
capitalized_list = []
for i in split_space:
    if i not in x_words:
        split_hyphen = i.split('-')
        temp_list = []
        for j in split_hyphen:
            temp_list.append(j[0].upper()+j[1:])
        capitalized_list.append('-'.join(temp_list))
    else:
        capitalized_list.append(i)    
capitalized = ' '.join(capitalized_list)
print('Capitalized: {}'.format(capitalized))

Enter a title: Welcome to the world of data-driven modeling for master students offered by UST in a great and green campus on the hill
Capitalized: Welcome to the World of Data-Driven Modeling for Master Students Offered by UST in a Great and Green Campus on the Hill


9. Draw a trigonometric table calculated for sin, cos and tan accurate to 3 decimal places in multiples of 10 degrees. All columns should be right aligned. Your output should look like this:

        deg	   sin	   cos	   tan  
          0	 0.000	 1.000	 0.000  
         30	 0.500	 0.866	 0.577
         60	 0.866	 0.500	 1.732
         90	 1.000	 0.000	 undef
        120	 0.866	-0.500	-1.732
        150	 0.500	-0.866	-0.577
        180	 0.000	-1.000	 0.000
        210	-0.500	-0.866	 0.577
        240	-0.866	-0.500	 1.732
        270	-1.000	 0.000	 undef
        300	-0.866	 0.500	-1.732
        330	-0.500	 0.866	-0.577
        360	 0.000	 1.000	 0.000

In [10]:
#Q9 answer
import math

tri_list = []
for i in range(0,361,30):
    temp_list = [i]
    if round(math.sin(math.radians(i)),3)==0:
        temp_list.append(0)
    else: temp_list.append(round(math.sin(math.radians(i)),3))
    if round(math.cos(math.radians(i)),3)==0:
        temp_list.append(0)
    else: temp_list.append(round(math.cos(math.radians(i)),3))
    if round(math.tan(math.radians(i)),3)==0:
        temp_list.append(0)
    elif abs(round(math.tan(math.radians(i)),3)) >10000:
        temp_list.append('undef')
    else: temp_list.append(round(math.tan(math.radians(i)),3))
    tri_list.append(temp_list)
        
print('{:>3}{:>9}{:>9}{:>9}'.format('deg','sin','cos','tan'))
for i in tri_list:
    try:
        print('{:>3}{:>9.3f}{:>9.3f}{:>9.3f}'.format(i[0],i[1],i[2],i[3]))
    except ValueError:
        print('{:>3}{:>9.3f}{:>9.3f}{:>9}'.format(i[0],i[1],i[2],i[3]))

deg      sin      cos      tan
  0    0.000    1.000    0.000
 30    0.500    0.866    0.577
 60    0.866    0.500    1.732
 90    1.000    0.000    undef
120    0.866   -0.500   -1.732
150    0.500   -0.866   -0.577
180    0.000   -1.000    0.000
210   -0.500   -0.866    0.577
240   -0.866   -0.500    1.732
270   -1.000    0.000    undef
300   -0.866    0.500   -1.732
330   -0.500    0.866   -0.577
360    0.000    1.000    0.000


10. List the annual extra profit amount up to cents with compount interest rate i.e. the extra interest made each year. Your output should look like this:

Enter a principal amount: 100  
Enter an annual interest rate (%): 10  
Enter the number of years: 5  

Year 1 profit: 10.00  
Year 2 profit: 11.00  
Year 3 profit: 12.10  
Year 4 profit: 13.31  
Year 5 profit: 14.64

In [11]:
#Q10 answer
principal = float(input('Enter a principal amount: '))
r = float(input('Enter an annual interest rate (%): '))
y = int(input('Enter the number of years: '))
print()

for i in range(1,y+1):
    interest = principal * r/100
    principal += interest
    print('Year {} profit: {:.2f}'.format(i,interest))
    

Enter a principal amount: 100
Enter an annual interest rate (%): 10
Enter the number of years: 5

Year 1 profit: 10.00
Year 2 profit: 11.00
Year 3 profit: 12.10
Year 4 profit: 13.31
Year 5 profit: 14.64


11. A quadratic equation $ax^2+bx+c=0$ can be solved by the quadratic formula $x=(-b±√(b²-4ac))/(2a)$. Each solution is a root of the quadratic function. Find out the 2 real roots, 1 real root or 2 complex roots of any given quadratic function. Your output should look like one of these:

Enter a quadratic function in ax^2+bx+c: x^2-3x+2  
The roots are 1 and 2  
***
Enter a quadratic function in ax^2+bx+c: 4x^2-4x+1  
The root is 0.5  
***
Enter a quadratic function in ax^2+bx+c: x^2-2x+2  
The roots are 1-i and 1+i

In [14]:
#Q11 answer
q = input('Enter a quadratic function in ax^2+bx+c: ')
position_x2 = q.find('x^2')
position_x = q.find('x',position_x2+1)

#find a
if position_x2==0:
    a = 1
else: a = int(q[:position_x2])

#find b
if position_x == -1:
    b = 0
else:
    b_str = q[position_x2+3:position_x]
    if len(b_str)==1:
        if b_str=='+':
            b = 1
        else: b = -1
    else: b = int(b_str)
        
#find c
if (len(q) == position_x2+3) or (len(q) == position_x+1):
    c=0
elif b==0:
    c = int(q[position_x2+3:])
else:
    c = int(q[position_x+1:])

#cal roots
import math
delta = b**2-4*a*c
if delta>0:
    s1 = (-b+math.sqrt(delta))/(2*a)
    s2 = (-b-math.sqrt(delta))/(2*a)
    if s1.is_integer():
        s1 = int(s1)
    if s2.is_integer():
        s2 = int(s2)
    print('The roots are {} and {}'.format(s2,s1))
elif delta==0:
    s = -b/(2*a)
    if s.is_integer():
        s = int(s)
    print('The root is {}'.format(s))
else:
    real = -b/(2*a)
    im = math.sqrt(-delta)/(2*a)
    if real.is_integer():
        real = int(real)
    if im.is_integer():
        im = int(abs(im))
    if im == 1:
        s1 = str(real)+'+i'
        s2 = str(real)+'-i'
    else:
        s1 = str(real)+'+'+str(im)+'i'
        s2 = str(real)+'-'+str(im)+'i'    
    print('The roots are {} and {}'.format(s2,s1))

Enter a quadratic function in ax^2+bx+c: x^2-2x+2
The roots are 1-i and 1+i
